src/main/java/org/spongepowered/common/world/FakePlayer.java

```diff
@@ -0,0 +1,194 @@
+import com.mojang.authlib.GameProfile;
+import net.minecraft.block.Block;
+import net.minecraft.block.material.Material;
+import net.minecraft.block.state.IBlockState;
+import net.minecraft.entity.player.EntityPlayerMP;
+import net.minecraft.item.Item;
+import net.minecraft.util.BlockPos;
+import net.minecraft.util.EnumFacing;
+import net.minecraft.util.MathHelper;
+import net.minecraft.world.WorldServer;
+import org.spongepowered.api.block.BlockState;
+import org.spongepowered.api.event.SpongeEventFactory;
+import org.spongepowered.api.event.block.InteractBlockEvent;
+import org.spongepowered.api.event.cause.Cause;
+import org.spongepowered.api.item.inventory.ItemStack;
+import org.spongepowered.api.util.Direction;
+import org.spongepowered.api.world.World;
+import org.spongepowered.common.SpongeImpl;
+import org.spongepowered.common.interfaces.world.IMixinWorld;
+import org.spongepowered.common.registry.provider.DirectionFacingProvider;
+import org.spongepowered.common.util.VecHelper;
+
+import java.util.Map;
+import java.util.Optional;
+import java.util.UUID;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+
+/**
+ * A marker to state that this is a fake player.
+ */
+public interface FakePlayer {
+
+    Controller controller = new Controller(SpongeImpl.getGame().createFakePlayerFactory());
+
+    public class Factory {
+
+        private final Map<UUID, EntityPlayerMP> players = new MapMaker().weakValues().makeMap();
+
+        public EntityPlayerMP getPlayer(WorldServer world, GameProfile profile) {
+            EntityPlayerMP player = players.get(profile.getId());
+            if (player == null) {
+                players.put(profile.getId(), player = new SimulatedPlayer(world, profile));
+            }
+            return player;
+        }
+
+    }
+
+    public static class Controller {
+
+        private final Factory factory;
+
+        private static Cause currentCause;
+        private static final Lock causeLock = new ReentrantLock();
+
+        private Controller(Factory factory) {
+            this.factory = factory;
+        }
+
+        public static Cause getCurrentCause() {
+            return currentCause;
+        }
+
+        private EntityPlayerMP setUpPlayer(GameProfile profile, World world, int x, int y, int z, ItemStack itemInHand, Cause cause) {
+            EntityPlayerMP player = factory.getPlayer((WorldServer) world, profile);
+            player.worldObj = (WorldServer) world;
+            player.theItemInWorldManager.setWorld((WorldServer) world);
+            player.posX = x;
+            player.posY = y;
+            player.posZ = z;
+            player.onGround = true;
+            player.inventory.setItemStack((net.minecraft.item.ItemStack) itemInHand);
+            player.inventory.mainInventory[player.inventory.currentItem] = (net.minecraft.item.ItemStack) itemInHand;
+
+            // Set up cause capture
+            causeLock.lock();
```

Comment:
I wasn't sure if putting a lock here was fine or not. I just don't want the `currentCause` to get overridden by anything while the simulation is processing.